File: SLR_Assignment_Deliverytime_sortingtime.R
#  Simple Linear Regression Assignment #
#  2) Delivery_time -> Predict delivery time using sorting time 
#  Do the necessary transformations for input variables for getting better R^2 value for the model prepared.

library(readr)
library(ggplot2)
library(moments)
DD_ST <- read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/delivery_time.csv")
View(DD_ST)
attach(DD_ST)
summary(DD_ST)
range(DD_ST$`Delivery Time`)
range(DD_ST$`Sorting Time`)
skewness(`Delivery Time`)
skewness(`Sorting Time`)

#Exploratory Data Analysis
boxplot(DD_ST$`Delivery Time`)
boxplot(DD_ST$`Sorting Time`)

#scatter plot for Caloriesconsumed vs Weightgained (Plot x,y)
plot(`Sorting Time`,`Delivery Time`)

#calculate correlation coefficient
cor(`Sorting Time`,`Delivery Time`)

#Simple Regression model
reg <- lm(`Delivery Time`~`Sorting Time`,data = DD_ST)
summary(reg)

#values prediction
#Confidence interval Calculation
confint(reg,level = 0.95)
pred <- predict(reg,interval = "predict")
#predict function gives fit value and its lower and upeer values as a range
pred <- as.data.frame(pred)
pred

#####Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x = `Sorting Time`, y = predict(reg, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')

cor(pred$fit,`Delivery Time`)

#Calculate Residuals "Errors"
reg$residuals
reg$residuals^2
mean(reg$residuals^2)
rmse <- sqrt(mean(reg$residuals^2))
rmse

############ Applying transformations##############
############ lOGORITHMIC MODEL    x = log(Sorting Time); y = Delivery Time ############
plot(log(`Sorting Time`),`Delivery Time`)
cor(log(`Sorting Time`),`Delivery Time`)

log_reg <- lm(`Delivery Time` ~ log(`Sorting Time`),data = DD_ST)
summary(log_reg)

#values prediction
#Confidence interval Calculation
confint(log_reg,level = 0.95)
pred_log <- predict(log_reg,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log <- as.data.frame(pred_log)
pred_log

rmse_log <- sqrt(mean(log_reg$residuals^2)) 
rmse_log
######or####
res_log=`Delivery Time`-pred_log$fit
rmse_log<-sqrt(mean(res_log^2))
rmse_log
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x =`Sorting Time`, y = predict(log_reg, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')

############ EXPONENTIAL MODEL   x = Sorting Time; y = log(Delivery Time) ############
plot(`Sorting Time`,log(`Delivery Time`))
cor(`Sorting Time`,log(`Delivery Time`))

log_reg2 <- lm(log(`Delivery Time`) ~ `Sorting Time`,data = DD_ST)
summary(log_reg2)

#values prediction
#Confidence interval Calculation
confint(log_reg2,level = 0.95)
pred_log2 <- predict(log_reg2,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log2 <- as.data.frame(pred_log2)

log_reg2$residuals #output is log(AT) so we are getting less values apply antilog
pred<- exp(pred_log2)  #anti-log=exponential
pred

cor(pred_log2$fit,`Delivery Time`)

res_log2=`Delivery Time`-pred$fit
rmse2 <- sqrt(mean(res_log2^2))
rmse2


##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x = `Sorting Time`, y = predict(log_reg2, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')

############Polynomial model with 2 degree (quadratic model)  ;x =Sorting Time^2 ; y = Delivery Time ############
#### input=x & X^2 (2-degree); output=y  ####
reg_quad2<- lm(`Delivery Time` ~ `Sorting Time`+I(`Sorting Time`*`Sorting Time`),data =DD_ST)
summary(reg_quad2)

#prediction
#Confidence interval Calculation
confint(reg_quad2,level = 0.95)
pred_quad2<-predict(reg_quad2,interval = "predict")
pred_quad2  <- as.data.frame(pred_quad2)
pred_quad2

resq=`Delivery Time`-pred_quad2$fit
rmse_quad<-sqrt(mean(resq^2))
rmse_quad


##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x = `Sorting Time`, y = predict(reg_quad2, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')

############Polynomial model with 3 degree (quadratic model)  ;x = Sorting Time^3; y = Delivery Time ############
#### input=x & X^2 & x^3 (3-degree); output=y  ####
reg_quad3<- lm(`Delivery Time` ~ `Sorting Time`+I(`Sorting Time`*`Sorting Time`)+I(`Sorting Time`*`Sorting Time`*`Sorting Time`),data =DD_ST)
summary(reg_quad3)

#prediction
#Confidence interval Calculation
confint(reg_quad3,level = 0.95)
pred_quad3<-predict(reg_quad3,interval = "predict")
pred_quad3  <- as.data.frame(pred_quad3)
pred_quad3

resq3=`Delivery Time`-pred_quad3$fit
rmse_quad3<-sqrt(mean(resq3^2))
rmse_quad3
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x = `Sorting Time`, y = predict(reg_quad3, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')






############ log transformation   x = log(Sorting Time); y = log(Delivery Time) ############
plot(log(`Sorting Time`),log(`Delivery Time`))
cor(log(`Sorting Time`),log(`Delivery Time`))

log_log_reg2 <- lm(log(`Delivery Time`) ~ log(`Sorting Time`),data = DD_ST)
summary(log_log_reg2)

#values prediction
#Confidence interval Calculation
confint(log_log_reg2,level = 0.95)
pred_log_log2 <- predict(log_log_reg2,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log_log2 <- as.data.frame(pred_log_log2)

log_log_reg2$residuals #output is log(AT) so we are getting less values apply antilog
pred<- exp(pred_log_log2)  #anti-log=exponential
pred

cor(pred_log_log2$fit,`Delivery Time`)

res_log_log2=`Delivery Time`-pred$fit
rmse_log2 <- sqrt(mean(res_log_log2^2))
rmse_log2


##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Sorting Time` , y =`Delivery Time` ),
             colour='red') + 
  geom_line(aes(x = `Sorting Time`, y = predict(log_log_reg2, newdata=DD_ST)),
            colour='blue') + 
  ggtitle('Sorting Time vs Delivery Time') +
  xlab('Sorting Time') +
  ylab('Delivery Time')

File: SLR_Assignment_Deliverytime_sortingtime.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 28 16:23:56 2020

@author: RAVI
"""
# For reading data set
# importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

DD_ST = pd.read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/delivery_time.csv")
DD_ST.columns="DeliveryTime","SortingTime"

import matplotlib.pylab as plt #for different types of plots

plt.scatter(x=DD_ST['SortingTime'], y=DD_ST['DeliveryTime'],color='green')# Scatter plot

np.corrcoef(DD_ST.SortingTime,DD_ST.DeliveryTime) #correlation
help(np.corrcoef)

import statsmodels.formula.api as smf
plt.hist(DD_ST["DeliveryTime"])

model = smf.ols('DeliveryTime ~ SortingTime', data=DD_ST).fit()
model.summary()

#values prediction
#Confidence interval Calculation
pred1 = model.predict(pd.DataFrame(DD_ST['SortingTime']))
pred1
print (model.conf_int(0.95)) # 95% confidence interval

res = DD_ST.DeliveryTime - pred1
sqres = res*res
mse = np.mean(sqres)
rmse = np.sqrt(mse)

######### Model building on Transformed Data#############

# Log Transformation
# x = log(SortingTime); y = DeliveryTime
plt.scatter(x=np.log(DD_ST['SortingTime']),y=DD_ST['DeliveryTime'],color='brown')
np.corrcoef(np.log(DD_ST.SortingTime), DD_ST.DeliveryTime) #correlation

model2 = smf.ols('DeliveryTime ~ np.log(SortingTime)',data=DD_ST).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(DD_ST['SortingTime']))
pred2
print(model2.conf_int(0.95)) # 95% confidence level

res2 = DD_ST.DeliveryTime - pred2
sqres2 = res2*res2
mse2 = np.mean(sqres2)
rmse2 = np.sqrt(mse2)

# Exponential transformation
plt.scatter(x=DD_ST['SortingTime'], y=np.log(DD_ST['DeliveryTime']),color='orange')

np.corrcoef(DD_ST.SortingTime, np.log(DD_ST.DeliveryTime)) #correlation

model3 = smf.ols('np.log(DeliveryTime) ~ SortingTime',data=DD_ST).fit()
model3.summary()

pred_log = model3.predict(pd.DataFrame(DD_ST['SortingTime']))
pred_log
pred3 = np.exp(pred_log)
pred3
print(model3.conf_int(0.95)) # 95% confidence level

res3 = DD_ST.DeliveryTime - pred3
sqres3 = res3*res3
mse3 = np.mean(sqres3)
rmse3 = np.sqrt(mse3)

############Polynomial model with 2 degree (quadratic model)  ;x = SortingTime*SortingTime; y = DeliveryTime############
#### input=x & X^2 (2-degree); output=y  ####
model4 = smf.ols('DeliveryTime ~ SortingTime+I(SortingTime*SortingTime)', data=DD_ST).fit()
model4.summary()

pred_p2 = model4.predict(pd.DataFrame(DD_ST['SortingTime']))
pred_p2

print(model3.conf_int(0.95)) # 95% confidence level

res4 = DD_ST.DeliveryTime - pred_p2
sqres4 = res4*res4
mse4 = np.mean(sqres4)
rmse4 = np.sqrt(mse4)

###########Polynomial model with 3 degree (quadratic model)  ;x = SortingTime*SortingTime*SortingTime; y = DeliveryTime############
#### input=x & X^2 (2-degree); output=y  ####
model5 = smf.ols('DeliveryTime ~ SortingTime+I(SortingTime*SortingTime)+I(SortingTime*SortingTime*SortingTime)', data=DD_ST).fit()
model5.summary()

pred_p3 = model5.predict(pd.DataFrame(DD_ST['SortingTime']))
pred_p3

print(model5.conf_int(0.95)) # 95% confidence level

res5 = DD_ST.DeliveryTime - pred_p3
sqres5 = res5*res5
mse5 = np.mean(sqres5)
rmse5 = np.sqrt(mse5)

# Log Transformation
# x = log(SortingTime); y = log(DeliveryTime)
plt.scatter(x=np.log(DD_ST['SortingTime']),y=np.log(DD_ST['DeliveryTime']),color='brown')
np.corrcoef(np.log(DD_ST.SortingTime), np.log(DD_ST.DeliveryTime)) #correlation

model6 = smf.ols('np.log(DeliveryTime) ~ np.log(SortingTime)',data=DD_ST).fit()
model6.summary()

pred_log6 = model6.predict(pd.DataFrame(DD_ST['SortingTime']))
pred_log6
pred6 = np.exp(pred_log6)
pred6
print(model6.conf_int(0.95)) # 95% confidence level

res6 = DD_ST.DeliveryTime - pred6
sqres6 = res6*res6
mse6 = np.mean(sqres6)
rmse6 = np.sqrt(mse6)


